File: 백준/Gold/1987. 알파벳/알파벳.py
import sys
input = sys.stdin.readline

from collections import deque

r, c = map(int, input().split())
graph = [list(input().rstrip()) for _ in range(r)]
queue = deque([(0,0,graph[0][0])])
dx = [0, 1, -1, 0]
dy = [1, 0, 0, -1]
answer = 0

while queue:
    x, y, s = queue.pop()
    answer = max(answer, len(s))

    for i in range(4):
        cur_x = x + dx[i]
        cur_y = y + dy[i]

        if cur_x < 0 or cur_x >= r:
            continue
        if cur_y < 0 or cur_y >= c:
            continue

        if graph[cur_x][cur_y] not in s:
            queue.append((cur_x, cur_y, s + graph[cur_x][cur_y]))
print(answer)

File: 백준/Gold/1987. 알파벳/알파벳.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt').toString().split('\n')

const [r, c] = input[0].split(' ').map(Number)
const board = []
for (let i = 0; i < r; i++) {
  board.push(input[i+1].split(''))
}

let answer = 0
const dx = [0, 1, 0, -1]
const dy = [1, 0, -1, 0]
const visited = new Set(board[0][0])
dfs(0, 0, 1)
console.log(answer)

function dfs(x, y, cnt) {
  answer = Math.max(answer, cnt)
  for (let i = 0; i < 4; i++) {
    const nx = dx[i] + x
    const ny = dy[i] + y
    if (nx < 0 || ny < 0 || nx >= r || ny >= c) continue

    const next_alphabet = board[nx][ny]
    if (visited.has(next_alphabet)) continue

    visited.add(next_alphabet)
    dfs(nx, ny, cnt + 1)
    visited.delete(next_alphabet)
  }
}